finetune/lora.py

```diff
@@ -124,7 +130,11 @@ def main(fabric: L.Fabric, data_dir: Path, checkpoint_dir: Path, out_dir: Path):
     fabric.print(f"Number of non trainable parameters: {num_parameters(model, requires_grad=False):,}")
     trainable_params = [p for p in model.parameters() if p.requires_grad]
 
-    optimizer = torch.optim.AdamW(trainable_params, lr=learning_rate, weight_decay=weight_decay)
+    if quantize and quantize.startswith("bnb."):
```

Comment:
A question: is paged optimizer tied to QLoRA only or it can be used anywhere?
Here is a quote from the paper:
> (3) Paged Optimizers, using NVIDIA unified memory to avoid the gradient checkpointing memory spikes that occur when processing a mini-batch with a long sequence length.

Here is one more:
> Paged Optimizers use the NVIDIA unified memory feature which does automatic page-to-page
transfers between the CPU and GPU for error-free GPU processing in the scenario where the GPU
occasionally runs out-of-memory. The feature works like regular memory paging between CPU RAM
and the disk. We use this feature to allocate paged memory for the optimizer states which are then
automatically evicted to CPU RAM when the GPU runs out-of-memory and paged back into GPU
memory when the memory is needed in the optimizer update step.

Reply:
Agree. But this task is not for me, especially multi-GPU part :)

Reply:
We currently only support QLoRA for 1 device at the moment (due to some other issues in bnb). But yeah, we can investigate wether we should use the Paged Optimizer in maybe in a separate PR.

Reply:
I tried the paged optimizer with Llama-2-7b-hf and nf4-dq, and I noticed a small difference between the runs with and without the paged optimizer.  The loss is exactly the same for the first few runs:

Paged:
iter 0 step 0: loss 12.6990, iter time: 448.06ms
iter 100 step 0: loss 13.0628, iter time: 164.10ms
iter 200 step 1: loss 12.6298, iter time: 167.23ms
iter 300 step 2: loss 12.6526, iter time: 149.93ms
iter 400 step 3: loss 12.4669, iter time: 163.87ms
iter 500 step 3: loss 12.9823, iter time: 145.37ms

non paged:
iter 0 step 0: loss 12.6990, iter time: 451.21ms
iter 100 step 0: loss 13.0628, iter time: 164.06ms
iter 200 step 1: loss 12.6298, iter time: 167.17ms
iter 300 step 2: loss 12.6526, iter time: 149.48ms
iter 400 step 3: loss 12.4659, iter time: 163.85ms
iter 500 step 3: loss 12.9822, iter time: 145.60ms

The difference becomes more pronounced as the runs progress:

Paged:
iter 10000 step 78: loss 7.5924, iter time: 129.82ms
iter 20000 step 156: loss 5.0993, iter time: 133.39ms
iter 30000 step 234: loss 6.7887, iter time: 142.82ms
iter 40000 step 312: loss 3.9812, iter time: 130.39ms
iter 49900 step 389: loss 6.4207, iter time: 143.68ms

non paged:
iter 10000 step 78: loss 7.5973, iter time: 129.92ms
iter 20000 step 156: loss 5.5243, iter time: 131.62ms
iter 30000 step 234: loss 6.7816, iter time: 144.36ms
iter 40000 step 312: loss 3.9866, iter time: 132.07ms
iter 49900 step 389: loss 6.3914, iter time: 142.37ms

The trends are the same in both runs, and the losses don't vary by more than about 0.5 across all the iterations (I only printed every 100).  Looks like a small rounding error or maybe even a floating-point optimization difference due to the nvcc compile-time options.

Reply:
Just tried the same with bf16-true and no quantization.  Similar results.  Loss started diverging by 300 iteration.

Still, appears like the paged optimizer works for non-quantized runs, at least for single GPUs.

Reply:
Thanks for the additional experiments! I am running a test matrix on the various setting atm and will report back!

Reply:
> I tried the paged optimizer with Llama-2-7b-hf and nf4-dq, and I noticed a small difference between the runs with and without the paged optimizer. The loss is exactly the same for the first few runs:

sorry i'm not following the work here closely but just wanted to point this out

https://twitter.com/Birchlabs/status/1687598048619192321